bridges/centralized-ethereum/src/actors/watch_dog.rs

```diff
@@ -0,0 +1,364 @@
+    pub start_ts: Option<Instant>,
+    /// Eth balance upon first metric report:
+    pub start_eth_balance: Option<f64>,
+    /// Wit balance upon last refund
+    pub start_wit_balance: Option<f64>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct WatchDogOutput {
+    pub running_secs: u64,
+}
+
+impl Drop for WatchDog {
+    fn drop(&mut self) {
+        log::trace!("Dropping WatchDog");
+        stop_system_if_panicking("WatchDog");
+    }
+}
+
+/// Make actor from EthPoller
+impl Actor for WatchDog {
+    /// Every actor has to provide execution Context in which it can run.
+    type Context = Context<Self>;
+
+    /// Method to be executed when the actor is started
+    fn started(&mut self, ctx: &mut Self::Context) {
+        log::debug!("WatchDog actor has been started!");
+
+        self.watch_global_status(None, None, ctx, Duration::from_millis(self.polling_rate_ms));
+    }
+}
+
+/// Required trait for being able to retrieve WatchDog address from system registry
+impl actix::Supervised for WatchDog {}
+impl SystemService for WatchDog {}
+
+impl WatchDog {
+    /// Initialize from config
+    pub fn from_config(config: &Config, eth_contract: Arc<Contract<Http>>) -> Self {
+        Self {
+            wit_jsonrpc_socket: config.witnet_jsonrpc_socket.to_string(),
+            wit_utxo_min_value_threshold: config.witnet_utxo_min_value_threshold,
+            eth_account: config.eth_from,
+            eth_contract: Some(eth_contract),
+            eth_jsonrpc_url: config.eth_jsonrpc_url.clone(),
+            polling_rate_ms: config.watch_dog_polling_rate_ms,
+            start_ts: Some(Instant::now()),
+            start_eth_balance: None,
+            start_wit_balance: None,
+        }
+    }
+
+    fn watch_global_status(
+        &mut self,
+        eth_balance: Option<f64>,
+        wit_balance: Option<f64>,
+        ctx: &mut Context<Self>,
+        period: Duration,
+    ) {
+        if self.start_eth_balance.is_none() && eth_balance.is_some() {
+            self.start_eth_balance = eth_balance;
+        }
+        if let Some(wit_balance) = wit_balance {
+            if wit_balance > self.start_wit_balance.unwrap_or_default() {
+                self.start_wit_balance = Some(wit_balance);
+                log::warn!("Wit account refunded to {} $WIT", wit_balance);
+            }
+        }
+        let start_eth_balance = self.start_eth_balance;
+        let start_wit_balance = self.start_wit_balance;
+        let wit_jsonrpc_socket = self.wit_jsonrpc_socket.clone();
+        let wit_utxo_min_value_threshold = self.wit_utxo_min_value_threshold;
+        let eth_jsonrpc_url = self.eth_jsonrpc_url.clone();
+        let eth_account = self.eth_account;
+        let eth_contract_address = self.eth_contract.clone().unwrap().address();
+        let running_secs = self.start_ts.unwrap().elapsed().as_secs();
+
+        let fut = async move {
+            let mut status = "up-and-running".to_string();
```

Comment:
What do you think about having all the possible values of status in an enum so they are easier to handle and less prone to error using them? Something like this

```rs
enum WatchDogStatus {
    // Add all of them here
    UpAndRunning,
    WitSyncing,
    Other(String),
}

impl WatchDogStatus {
    fn to_string(&self) -> String {
        match self {
            WatchDogStatus::UpAndRunning => "up-and-running".to_string(),
            WatchDogStatus::WitSyncing => "wit-syncing".to_string(),
            WatchDogStatus::Other(message) => message.clone(),
        }
    }
}
```

Reply:
no time before holidays, next version perhaps